src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,457 @@
+  }
+
+  def saturate(data: Option[Config.ProverStateSaturationTimeout]): Unit = {
+    endPreamblePhase()
+    data match {
+      case Some(Config.ProverStateSaturationTimeout(timeout, comment)) => saturate(timeout, comment)
+      case None => /* Don't do anything */
+    }
+  }
+
+  def saturate(timeout: Int, comment: String): Unit = {
+    endPreamblePhase()
+    setTimeout(Some(timeout))
+    prover.check()
+  }
+
+  protected def retrieveAndSaveModel(): Unit = {
+    if (Verifier.config.counterexample.toOption.isDefined) {
+      val model = prover.getModel
+      lastModel = model
+    }
+  }
+
+  protected def assertUsingSoftConstraints(goal: Term): (Boolean, Long) = {
+    endPreamblePhase()
+    val guard = fresh("grd", Nil, sorts.Bool)
+    val guardApp = App(guard, Nil)
+    val goalImplication = Implies(guardApp, goal)
+
+    prover.add(termConverter.convertTerm(goalImplication).asInstanceOf[BoolExpr])
+
+    val startTime = System.currentTimeMillis()
+    val res = prover.check(termConverter.convertTerm(guardApp))
+    val endTime = System.currentTimeMillis()
+    val result = res == Status.UNSATISFIABLE
+    if (!result) {
+      retrieveAndSaveModel()
+    }
+
+    (result, endTime - startTime)
+  }
+
+  def check(timeout: Option[Int] = None): Result = {
+    endPreamblePhase()
+    setTimeout(timeout)
+
+    val res = prover.check()
+
+    res match {
+      case Status.SATISFIABLE => Sat
+      case Status.UNSATISFIABLE => Unsat
+      case Status.UNKNOWN=> Unknown
+    }
+  }
+
+  def endPreamblePhase() =  {
+    if (!preamblePhaseOver) {
+      preamblePhaseOver = true
+
+      val merged = emittedPreambleString.mkString("\n")
+      val parsed = ctx.parseSMTLIB2String(merged, emittedSortSymbols.toArray, emittedSorts.toArray, emittedFuncSymbols.toArray, emittedFuncs.toArray)
+      prover.add(parsed: _*)
+      prover.add(preambleAssumes.toSeq : _*)
+      preambleAssumes.clear()
+    }
+  }
+
+  def statistics(): Map[String, String] = {
+    val statistics = prover.getStatistics
+    val result = mutable.HashMap[String, String]()
+    for (e <- statistics.getEntries()) {
+      result.update(e.Key, e.getValueString)
+    }
+    ListMap.from(result)
+  }
+
+  def comment(str: String): Unit = {
+    // ignore
+  }
+
+  def fresh(name: String, argSorts: Seq[Sort], resultSort: Sort): Fun = {
+    val id = identifierFactory.fresh(name)
+    val fun = Fun(id, argSorts, resultSort)
+    val decl = FunctionDecl(fun)
+
+    termConverter.convert(decl)
+
+    fun
+  }
+
+  def declare(decl: Decl): Unit = {
+    // We convert the declaration into a Z3-level declaration (which is automatically added to Z3's
+    // current state) and record it in out collection(s) of emmitted declarations.
+    // Special handling for macro declarations if expandMacros is true; in that case,
+    // nothing is declared on the Z3 level, and we simply remember the definition ourselves.
+    decl match {
+      case SortDecl(s) => {
```

Comment:
Braces around case-bodies are not necessary, and IntelliJ suggests to remove them. Please do that.